### PR TITLE
Fixture continuous capabilities

### PR DESCRIPTION
**Pull request by hjtappe**
Inspired by #1370, I remembered the deactivated check in the fixtures-tool and got the ambitious idea to fix the fixtures to be able to activate that check for non-contiguous capability ranges.

Based on the manuals I found on the internet (or on arithmetics), I fixed the fixtures to pass the check by adding capabilities, capability ranges and "No function" capabilities.

While going along, I noticed quite a few "No function" entries. As a first step, I unified their spelling throughout the fixtures. In a future step, we might want to add a Preset "NoFunction" for the capabilities as we currently have it in the channels presets. It is more or less a in-place sed commandline to proceed with this (minus "No function" entries which already have some Preset already - which in some cases is not valid and could be fixed). I refrained from adding this due to the sync with the Open Fixture Library repository which has this Preset already on their side.
Adding the "NoFunction" Preset would allow to hide or grey out these entries (except for the first one "min=0") in the qmlui if you like.

Also, as going through the fixtures, I activated the "mode name check" in the fixtures-tool for those fixtures being added recently. Updating that for the older fixtures poses a breaking change, which we might do with end of QLC+ 4.x / start of only QLC+ 5.x or never - or we let the importer search for a "... Mode" mode while importing to compensate this.

Please review this pull request and let me know of required changes or reverts.

### CODE REVIEW

**Review by mcallegari — CHANGES_REQUESTED** [review]
I haven't checked every single file, but my comments should give you an idea for the whole PR.
Some "No function" are invented fillers and should be checked against manuals.
I would remove the "mode" check or check it against version 4.12.7.
It is very difficult to review a large PR like this. It takes hours.
Maybe split it in batches of 100 files maximum?

File: resources/fixtures/American_DJ/American-DJ-Inno-Pocket-Spot.qxf [hunk 1]
```diff
@@ -80,14 +81,14 @@
  </Channel>
  <Channel Name="Dimmer Curves">
   <Group Byte="0">Maintenance</Group>
-  <Capability Min="0" Max="0">Standard</Capability>
+  <Capability Min="0" Max="41">Standard</Capability>
```
[the remark this answers is not in the document]
Comment (hjtappe) [reply resources/fixtures/American_DJ/American-DJ-Inno-Pocket-Spot.qxf:84]: According to your link, the whole block is different from the manual I found: http://adjmedia.s3-website-eu-west-1.amazonaws.com/manuals/2%20Kopie%20von%20Inno%20Pocket%20Spot_02_GER.pdf
There might be two version (MK I and MK II) of that fixture?

I have corrected the block now to the manual you provided.

In 98%, I went through the manuals that were available on the internet to make up the corrections. Only some fixtures were so old that no manuals were provided. ADJ normally had all manuals available as far as I remember.

File: resources/fixtures/American_DJ/American-DJ-Quad-Gem-DMX.qxf [hunk 2]
```diff
@@ -18,10 +18,12 @@
  <Channel Name="Mode Pattern">
   <Group Byte="0">Maintenance</Group>
   <Capability Min="0" Max="119">Pattern</Capability>
+  <Capability Min="120" Max="239">No function</Capability>
```
[the remark this answers is not in the document]
Comment (hjtappe) [reply resources/fixtures/American_DJ/American-DJ-Quad-Gem-DMX.qxf:21]: I agree. Not sure how I came up with "no function" here.
http://adjmedia.s3-website-eu-west-1.amazonaws.com/manuals/D_Quad_Gem_DMX.pdf

File: resources/fixtures/American_DJ/American-DJ-Inno-Pocket-Spot.qxf [hunk 3]
```diff
@@ -80,14 +81,14 @@
...
   <Capability Min="42" Max="84">Stage</Capability>
   <Capability Min="85" Max="127">TV</Capability>
   <Capability Min="128" Max="170">Architectural</Capability>
   <Capability Min="171" Max="213">Theatre</Capability>
   <Capability Min="214" Max="255">Default to unit curve setting</Capability>
  </Channel>
- <Mode Name="9 Channel Mode">
+ <Mode Name="9 Channel">
```
[the remark this answers is not in the document]
Comment (hjtappe) [reply resources/fixtures/American_DJ/American-DJ-Inno-Pocket-Spot.qxf:91]: OK, reverted now. It was only four fixtures since 4.11.0 having this problem and being included in 4.12.6.

File: resources/fixtures/American_DJ/American-DJ-Accu-Spot-250-II.qxf [hunk 4]
```diff
@@ -91,9 +91,9 @@
   <Capability Min="0" Max="225">Max -&gt; Min</Capability>
   <Capability Min="226" Max="235">Blackout By Movement</Capability>
   <Capability Min="236" Max="245">Blackout By Wheel Movement</Capability>
-  <Capability Min="246" Max="255">No Function</Capability>
+  <Capability Min="246" Max="255">No function</Capability>
  </Channel>
- <Mode Name="8 Channel Mode">
+ <Mode Name="8 Channel">
```
[the remark this answers is not in the document]
Comment (hjtappe) [reply resources/fixtures/American_DJ/American-DJ-Accu-Spot-250-II.qxf:96]: OK, reverted.

File: resources/fixtures/American_DJ/American-DJ-Inno-Scan-LED.qxf [hunk 5]
```diff
@@ -90,7 +90,7 @@
   <Capability Min="194" Max="255" Preset="RotationCounterClockwiseSlowToFast">Counter - Clockwise rotation Slow to Fast</Capability>
  </Channel>
  <Channel Name="Pan / Tilt Speed" Preset="SpeedPanTiltSlowFast"/>
- <Mode Name="8 Channel Mode">
+ <Mode Name="8 Channel">
```
[the remark this answers is not in the document]
Comment (hjtappe) [reply resources/fixtures/American_DJ/American-DJ-Inno-Scan-LED.qxf:93]: OK reverted.

File: resources/fixtures/American_DJ/American-DJ-Entour-Faze.qxf [hunk 6]
```diff
@@ -26,6 +26,7 @@
  <Channel Name="Interval">
   <Group Byte="0">Speed</Group>
   <Capability Min="0" Max="1">0 Seconds</Capability>
+  <Capability Min="2" Max="8">No function</Capability>
```
[the remark this answers is not in the document]
Comment (hjtappe) [reply resources/fixtures/American_DJ/American-DJ-Entour-Faze.qxf:29]: Your values are for duration, this is the interval capability with
0-1: 0s
9-255: 5-255s

In this case, there is a gap in the manual for 2-8, which effectively is a "reserved / no function" in my perception and could be the same as "no interval (0s)". It could as well mean "2-9: 1-4s" and be forgotten in the manual. The real behavior can probably only be resolved with the real device at hand.
Thus the proposal to fill this with a "No Function" as long as no-one complains.

**Comment by hjtappe** [comment]
@mcallegari, i bear with you regarding the review process. It has been ~800 places, as already indicated by the comment, taking days to find the manuals and deriving the solutions from them. Should I split this into smaller chunks:

- mode name check
- unification of "No function" spelling (to provide adding this as Capability Preset if useful in the UI for greying out)
- Fixes, split by alphanumerical groups

Give this comment a thumbs-up and I will start that process.

File: resources/fixtures/American_DJ/American-DJ-Entour-Faze.qxf (hunk 6 above)
[the remark this answers is not in the document]
Comment (mcallegari) [reply resources/fixtures/American_DJ/American-DJ-Entour-Faze.qxf:29]: Ok

File: resources/fixtures/Robe/Robe-Robin-300E-Beam.qxf [hunk 7]
```diff
@@ -111,33 +119,42 @@
...
-  <Capability Min="65" Max="70" Preset="GoboMacro" Res1="SGM/gobo00061.svg">Positioning beam reducer (5.5°)</Capability>
-  <Capability Min="71" Max="75" Preset="GoboMacro" Res1="SGM/gobo00062.svg">Positioning beam reducer (4.5°)</Capability>
-  <Capability Min="76" Max="81" Preset="GoboMacro" Res1="SGM/gobo00063.svg">Positioning beam reducer (3°)</Capability>
-  <Capability Min="82" Max="86" Preset="GoboMacro" Res1="SGM/gobo00064.svg">Positioning beam reducer (1.5°)</Capability>
-  <Capability Min="87" Max="92" Preset="GoboMacro" Res1="Robe/gobo00070.svg">Positioning 1</Capability>
-  <Capability Min="93" Max="97" Preset="GoboMacro" Res1="Robe/gobo00027.svg">Positioning 2</Capability>
-  <Capability Min="98" Max="103" Preset="GoboMacro" Res1="Robe/gobo00026.svg">Positioning 3</Capability>
-  <Capability Min="104" Max="109" Preset="GoboMacro" Res1="Robe/gobo00025.svg">Positioning 4</Capability>
+  <Capability Min="0" Max="0" Preset="GoboMacro" Res1="Others/open.svg">Open/hole (6.5°)</Capability>
+  <Capability Min="1" Max="6">Open/Beam reducer (5.5)</Capability>
```
Comment (mcallegari) [inline resources/fixtures/Robe/Robe-Robin-300E-Beam.qxf:123]: I'm not really sure how this could be treated.
Intermediate positions are quite useless in most cases.
I think this should be "0-6" open, "7-13" reducer and so on.
Then QLC+ should handle the intermediate positions

File: resources/fixtures/Robe/Robe-Robin-600E-Beam.qxf [hunk 8]
```diff
@@ -113,19 +121,27 @@
  <Channel Name="Static Gobos">
   <Group Byte="0">Gobo</Group>
   <Capability Min="0" Max="0">Open</Capability>
+  <Capability Min="1" Max="6">Open/Gobo 1</Capability>
```
Comment (mcallegari) [inline resources/fixtures/Robe/Robe-Robin-600E-Beam.qxf:124]: Same as above

File: resources/fixtures/Robe/Robe-Robin-600E-Spot.qxf [hunk 9]
```diff
@@ -116,7 +124,27 @@
  </Channel>
  <Channel Name="Static Gobos">
   <Group Byte="0">Gobo</Group>
-  <Capability Min="0" Max="0">Open</Capability>
+  <Capability Min="0" Max="0" Preset="GoboMacro" Res1="Others/open.svg">Open/hole</Capability>
+  <Capability Min="1" Max="6">Open/Gobo 1</Capability>
```
Comment (mcallegari) [inline resources/fixtures/Robe/Robe-Robin-600E-Spot.qxf:128]: Same as above

File: resources/fixtures/Robe/Robe-Robin-DLX-Spot.qxf [hunk 10]
```diff
@@ -87,15 +87,25 @@
...
-  <Capability Min="7" Max="7">Gobo 1 - continual positioning</Capability>
-  <Capability Min="13" Max="13">Gobo 2 - continual positioning</Capability>
-  <Capability Min="19" Max="19">Gobo 3 - continual positioning</Capability>
-  <Capability Min="26" Max="26">Gobo 4 - continual positioning</Capability>
-  <Capability Min="32" Max="32">Gobo 5 - continual positioning</Capability>
-  <Capability Min="39" Max="39">Gobo 6 - continual positioning</Capability>
-  <Capability Min="45" Max="45">Gobo 7 - continual positioning</Capability>
-  <Capability Min="51" Max="51">Gobo 8 - continual positioning</Capability>
-  <Capability Min="58" Max="58">Gobo 9 - continual positioning</Capability>
+  <Capability Min="1" Max="6">Open/Gobo 2</Capability>
```
Comment (mcallegari) [inline resources/fixtures/Robe/Robe-Robin-DLX-Spot.qxf:90]: Same as above

**Review by mcallegari — CHANGES_REQUESTED** [review]

**Comment by mcallegari** [comment]
At last I reviewed all of it. Took an awful amount of time because of brain context switching.
In the future, please split PRs in topics. This PR could be split at least into 3 PRs like:
PR1 -> "No Function" to "No function"
PR2 -> fill capability gaps
PR3 -> improve existing capabilities

Anyway, thanks for the huge work.
Let's make the fixes for my latest comments and merge this.

**Comment by mcallegari** [comment]
Also, please search for "funktion" (Cameo Storm) and "Fonction" (PR 5000 Spot)

File: resources/fixtures/JB_Systems/JB-Systems-The-WinnerII.qxf [hunk 11]
```diff
@@ -19,37 +19,65 @@
  <Channel Name="Gobo">
   <Group Byte="0">Gobo</Group>
   <Capability Min="0" Max="0" Preset="ColorMacro" Res1="#000000">Blackout</Capability>
+  <Capability Min="1" Max="7">Blackout --&gt; Open</Capability>
   <Capability Min="8" Max="8" Preset="GoboMacro" Res1="Others/open.svg">Open</Capability>
+  <Capability Min="9" Max="15">Open --&gt; Circle</Capability>
   <Capability Min="16" Max="16" Preset="GoboMacro" Res1="SGM/gobo00063.svg">1 - Circle</Capability>
+  <Capability Min="17" Max="23">Circle --&gt; Square outline</Capability>
```
[the remark this answers is not in the document]
Comment (hjtappe) [reply resources/fixtures/JB_Systems/JB-Systems-The-WinnerII.qxf:26]: @mcallegari, in many (but not all) fixtures with Gobos, there are the fixed Gobo positions as dedicated Capability, in some there are not. I did not start unifying these where that would not solve the problems of capabilities missing or messed up.

In this case, we have dedicated positions for the full gobo positions, including the GoboMacro and intermediate positions. It makes sense to me because a) in the VC, these will be rendered in two columns, the first column being the full positions and the second the intermediate positions and b) for the full positions, I do not have to take care to reach the lowest value (as the VC widget precision depends on the position where the button receives the click release event, so I could unintentionally select an intermediate position if not clicking carefully enough).

I would not know which intermediate Gobo SVG file I would chose for these intermediate positions. The group type "Gobo" should be sufficient for the program to handle this correctly. What else would the Preset=GoboMacro achieve (without the SVG file)?

Please advise.

File: resources/fixtures/Robe/Robe-Robin-300E-Beam.qxf (hunk 7 above)
Comment (hjtappe) [reply resources/fixtures/Robe/Robe-Robin-300E-Beam.qxf:123]: For intermediate color positions, I see the benefit, but for the intermediate gobo and beam reducer positions, I also see no use for them (they say fine positioning - so moving a cropped beam around in the full beam window), except maybe for very rare and special cases. Nevertheless, the manufacturers provided these as a separate feature (the fixed positions with DMX ranges are there, too), so I think, we will need to represent them.
https://www.robe.cz/res/downloads/dmx_charts/Robin_300E_Beam_DMX_charts.pdf

If so, we would need to make space around the full positions, but I wouldn't go so far:
0-4: Open 6.5°
5-12: Reducer 5.5°
...
61-64: Open 6.5°

File: resources/fixtures/Clay_Paky/Clay-Paky-Tambora-Flash.qxf [hunk 12]
```diff
@@ -516,7 +516,7 @@
   <Channel Number="34">Layer 3 Strobe Indexing Speed Nothing</Channel>
   <Channel Number="35">Layer 3 Strobe Fade</Channel>
  </Mode>
- <Mode Name="Pixels RGB mode">
+ <Mode Name="Pixels RGB">
```
[the remark this answers is not in the document]
Comment (hjtappe) [reply resources/fixtures/Clay_Paky/Clay-Paky-Tambora-Flash.qxf:519]: This was not in 4.12.6 and indicates the 4.12.7 version, so it should not be a regression. But reverting will make the check trigger the mode error.
I would propose to leave it without the "mode" in the mode names, especially as only three of many other modes have that in the name.

File: resources/fixtures/Robe/Robe-Robin-300E-Beam.qxf (hunk 7 above)
Comment (mcallegari) [reply resources/fixtures/Robe/Robe-Robin-300E-Beam.qxf:123]: Same as my comment above. Let's leave these without preset for now.

**Review by mcallegari — APPROVED** [review]
I think this is OK now and can be merged.
Shall I proceed?

**Comment by hjtappe** [comment]
> I think this is OK now and can be merged. Shall I proceed?

From my point of view yes, this PR can be merged.

There are still some improvable things in the fixtures, that I thought of, but I need your view on this and they would be worth another PR if we want to address them:

1) We might introduce a Preset="NoFunction" for Capabilities. This is also present in the open lighing project, where the converter would also need to be adapted. This would allow us to hide or grey out the fillter "No function" Capabilities unless they start at zero (default entries) or is at a shifted default DMX address for that channel.

2) There is a huge inconsistency between color codes and color names, not just small adaptions to fixture specifics. I was looking for the ColorMacro color codes where I had to add them and used this script which I later removed: https://github.com/mcallegari/qlcplus/commit/5f1f5cdd785d63d5db522f096ff300d12cc468d1#diff-07e4c1317f06cde6fd40efcbf104e112e19242e04b5101656716998856705e14
Running for example `./resources/fixtures/scripts/getFixtureColorNames.sh | grep Cyan` provides you with color codes for deep blue, clear green and magenta within different codes which are more like Cyan. We could use a list like https://htmlcolorcodes.com/color-names/ and add another check to keep the color codes unique for the given names.

Let me know if this sounds interesting or not worth it on the long run.

**Comment by mcallegari** [comment]
> 1. We might introduce a Preset="NoFunction" for Capabilities. This is also present in the open lighing project, where the converter would also need to be adapted. This would allow us to hide or grey out the fillter "No function" Capabilities unless they start at zero (default entries) or is at a shifted default DMX address for that channel.

Agree. This could be useful.

> 2. There is a huge inconsistency between color codes and color names, not just small adaptions to fixture specifics. I was looking for the ColorMacro color codes where I had to add them and used this script which I later removed: [5f1f5cd#diff-07e4c1317f06cde6fd40efcbf104e112e19242e04b5101656716998856705e14](https://github.com/mcallegari/qlcplus/commit/5f1f5cdd785d63d5db522f096ff300d12cc468d1#diff-07e4c1317f06cde6fd40efcbf104e112e19242e04b5101656716998856705e14)
>    Running for example `./resources/fixtures/scripts/getFixtureColorNames.sh | grep Cyan` provides you with color codes for deep blue, clear green and magenta within different codes which are more like Cyan. We could use a list like https://htmlcolorcodes.com/color-names/ and add another check to keep the color codes unique for the given names.

Yes, this is another Pandora box of fixture definitions :smile: 
In most of the cases, ColorMacro RGB values don't even represent the real color produced by a fixture. There are so many variables like color temperature, CRI, HCR, etc
Let's say QLC+ gives you the idea of the color...not quite the exact color.
But I do agree with you they should be standardized. What you referenced is quite similar to
https://github.com/mcallegari/qlcplus/blob/master/resources/colorfilters/namedrgb.qxcf
The best option would be to match the string the user enters in the capability with those colors, and set the ColorMacro automatically.
Unfortunately there are guys like LEE and Pantone who didn't give their OK to the QLC+ project (as if some RGB values can damage them...also considering those values are public on their websites...)
In conclusion, it would be "easy" to fix named colors but not "proprietary" filters.
I would leave this topic open for the future but I guess there won't ever be an easy patch for the problem.

For now, let's merge this. Thanks again for the huge amount of time you dedicated to this PR!
